### ISSUE TITLE
Arcface model from onnx model zoo returns nan with CUDA

### ISSUE THREAD

**Issue by mika-fischer** [microsoft/onnxruntime#831]
**Describe the bug**
When using the model from here:
https://github.com/onnx/models/tree/master/models/face_recognition/ArcFace

inference with onnxruntime-gpu 0.3.0 (and also using the C-API with current master) returns all nans. With onnxruntime (non-GPU), the results are as expected.

**System information**
Happens on Linux and Windows, with 0.3.0 and current master.

**To Reproduce**
* Download https://s3.amazonaws.com/onnx-model-zoo/arcface/resnet100/resnet100.onnx
* Then use this test script
```python
import sys, numpy, onnxruntime 
sess = onnxruntime.InferenceSession(sys.argv[1])
feeds = {i.name: numpy.zeros(shape=[d if (d and d > 0) else 1 for d in i.shape], dtype=numpy.float32)
         for i in sess.get_inputs()}
fetches = [o.name for o in sess.get_outputs()]
results = sess.run(fetches, feeds)
for name, val in zip(fetches, results):
    print(f'{name} ({val.shape}):')
    print(val)
```

This results in:
```
fc1 ((1, 512)):
[[nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
  nan nan nan nan nan nan nan nan]]
```

**Comment by hariharans29** [comment]
Hi @mika-fischer,

Are you noticing the first nans in the node "bn0" (the first batchnorm operation) ? As a first step to quickly repro - I wrote a test case that simulated running this specific node alone - luckily scale, B, mean. var all seem to be initializers - so I took their values directly from the model file. They are all having shape - [64], so I gave it arbitrary input X of shape [1, 64, 1, 1]  (i.e.) 64 values and the test passed for gpu.

Let me try running the full model.

**Comment by mika-fischer** [comment]
By now I figured out that the model works correctly if you change the "spatial" attribute of all BatchNormalization nodes from 0 to 1.

However, I'm not really sure why that helps. Maybe the cudnn computations are more numerically stable in that mode...

**Comment by mika-fischer** [comment]
Oh, and I think the numerical issue comes from the fact that the first conversation layer had some filters that are very close to zero. So the activations are also very close to zero. Therefore the BatchNormalization variance is also very close to zero. And therefore the BatchNormalization computation runs into numerical issues.

**Comment by hariharans29** [comment]
Hi @mika-fischer ,

Thanks for your reply. 

In my debugging I could see that "bn0" (the very first BN node) produces the first NaNs (as you indicated). But here is where it gets strange - since for BN the output shape is the same as the input shape, I tried to take a corresponding input to an output value which was NaN. One such input (output of the convolutional layer) was -0.001681. 

Next, I tried to write a "single node" test case to repro this - I copied over the initializer values for mean/var/scale/bias - all 1D tensors of shape [64]. I created a "fake" input tensor of shape [1, 64, 1, 1] all filled with  -0.001681 (spread across all channels - so as to trigger the NaN creation in atleast one of them). I expect atleast one of the output values to be a NaN. It is not and it produces correct results (both for spatial = 1 and spatial = 0).

**Comment by mika-fischer** [comment]
Yes, something strange is going on, which I don't yet understand. Note that the BatchNorm as used in this model is spatial, i.e. in the model the input dimensions are never [1, 64, 1, 1] but rather [1, 64, H, W] and the batchnorm uses broadcasting to apply the same transformation to all locations. This is spatial mode, so it would make sense that the model would use spatial = 1 in the first place.

Also do note that onnxruntime uses a different cudnn batchnorm mode for spatial == 1 and spatial == 0 (see https://github.com/Microsoft/onnxruntime/blob/7e1edbb9a2cc0dce75940fc7e5695a852a2337e2/onnxruntime/core/providers/cuda/nn/batch_norm.h#L28-L30).

However, I still don't understand what's actually happening with spatial = 0. I had actually assumed that that mode would not even work, since I assumed that cudnn then expects the dimensions of input, output, mean, variance, etc. to all match exactly (except the batch size), which is not the case in this model.

But it does in fact accept the values, just gives different (wrong) results. Does it broadcast also? Does it perform different calculations with more numerical issues? I have no idea...

What's also strange is that the resnet34 models (just as an example) have basically the same structure and also specify spatial = 0. But they do in fact work correctly!

Thanks for taking a look at this! I'll also take another look on Tuesday when I'm back at the office.

**Comment by hariharans29** [comment]
Hi @mika-fischer,

1) True, the model input dimensions are never [1, 64, 1, 1] but rather [1, 64, H, W], but the idea was to find an input value amongst the (1 * 64 * H * W) values that leads to a NaN output value and spread it out across all channels and try and repro the NaN generation by using a single BN node (and other initializers as provided in the model).

As far as I see (I could be wrong - please correct me if I am), there doesn't seem to be any mention of broadcasting in the BN spec. AFAIK, BN simply does this per channel -
   output_value = scale * ((input_value - channel_mean) / sqrt(channel_var + epsilon)) + shift

Hence, my decision to spread the value  -0.001681 across all the 64 channels to get a repro.

2) Please refer to PR #890 - the CPU provider does not support spatial != 1 in it's implementation and hence should never have loaded the model - this is now fixed. So, this model as is should never have been able to run using the CPU package. When modified to spatial == 1, it will run.

3) Also, do you know why the model has BN nodes with spatial == 0 when the results are correct even when they are modified to spatial == 1 ? 

Thanks!

**Comment by mika-fischer** [comment]
Hi, @hariharans29,

Thanks for getting back to me and sorry for not being able to continue my investigation (I'll probably not find enough time to debug this in the next few weeks either...).

Quoting from https://github.com/onnx/onnx/blob/master/docs/Changelog.md#batchnormalization-7:
> Input data tensor from the previous operator; dimensions for image case are (N x C x H x W), where N is the batch size, C is the number of channels, and H and W are the height and the width of the data. For non image case, the dimensions are in the form of (N x C x D1 x D2 ... Dn), where N is the batch size.

> If spatial is true, the dimension of the running mean (training) or the estimated mean (testing) is (C). If spatial is false, the dimensions of the running mean (training) or the estimated mean (testing) are (C x D1 x ... x Dn).

And similarly for the variance.

Conversely this means that if you have an input shape of [1, 64, 1, 1] the mean present in the model has the correct number of elements for both spatial == 1 and spatial == 0, albeit for spatial == 0, the shape is incorrect (should be [64, 1, 1] but is [64]).

Whereas with an input shape of [1, 64, H, W] with H>1 and W>1, the mean and variance in the model don't have the right shape *and* don't have the correct number of elements.

Apparently in the onnx spec, broadcasting is always explicitly mentioned, and it isn't mentioned for the BatchNorm operator. So it is probably a bug that onnxruntime even accepts the model with spatial == 0 with the CUDA provider...

But nevertheless, there is another issue:
At the bottom of it, the only difference between spatial == 0 and spatial == 1 I can see in the CUDA provider is whether the cudnn Tensors and cudnn BatchNorm function are called with CUDNN_BATCHNORM_SPATIAL or CUDNN_BATCHNORM_PER_ACTIVATION.

Quoting https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#cudnnBatchNormMode_t:
> CUDNN_BATCHNORM_PER_ACTIVATION
Normalization is performed per-activation. This mode is intended to be used after non-convolutional network layers. In this mode the tensor dimensions of bnBias and bnScale, the parameters used in the cudnnBatchNormalization* functions, are 1xCxHxW.

> CUDNN_BATCHNORM_SPATIAL
Normalization is performed over N+spatial dimensions. This mode is intended for use after convolutional layers (where spatial invariance is desired). In this mode the bnBias, bnScale tensor dimensions are 1xCx1x1.

So, also for cudnn the model in question should not work with spatial == 0, because the mean and variance in the model have the wrong shape and wrong number of elements. _And yet it does work correctly_ (for instance for the resnet models https://github.com/onnx/models/tree/master/models/image_classification/resnet).

So I assume that either onnxruntime broadcasts the shape of the mean and variance so that it matches what cudnn expects, or cudnn itself does this...

If in fact broadcasting is happening somewhere, then the two cases spatial == 0 and spatial == 1 should do exactly the same thing.

And yet with this model, they don't. I can only assume that the two cudnn BatchNorm implementations don't do the same thing internally (i.e. have numerical differences).

I hope this helps you track down the remaining issues. If you have more questions please feel free to ask.

Regarding your commit: Might I suggest not just disabeling the tests for this model, but actually fixing the model (by setting spatial = 1), and also forward this problem and the fix to the onnx model zoo, where basically every model exported from mxnet will have this issue. They won't work with the CPU provider after your commit and they don't work reliably with the CUDA provider, either...

**Comment by snnn** [comment]
Updated the title because the model doesn't work on CPU.

**Comment by faxu** [comment]
https://github.com/onnx/models/issues/156

**Comment by luan1412167** [comment]
Hi, I'm facing such the error. Can't you help me pointing out where I can modify spatial for BatchNorm. 
`2019-10-07 11:54:01.026045667 [E:onnxruntime:test, inference_session.cc:547 Initialize] Exception during initialization: /onnxruntime_src/onnxruntime/core/providers/cpu/nn/batch_norm.h:39 onnxruntime::BatchNorm<T>::BatchNorm(const onnxruntime::OpKernelInfo&) [with T = float] spatial == 1 was false. BatchNormalization kernel for CPU provider does not support non-spatial cases
`

**Comment by snnn** [comment]
Hi @luan1412167  Sorry, onnxruntime doesn't support this model yet.

**Comment by luan1412167** [comment]
@snnn Have you procedure for fixing it?. Can I run it on GPU?

**Comment by hariharans29** [comment]
Hi @mika-fischer,

We finally resolved why the model produces NaNs in the CUDA provider. This is because the **ONNX equivalent** of the original model is **invalid**. We have requested an update to the model here -  https://github.com/onnx/models/issues/156. Please see my reply in the thread wherein I detail why it is invalid.

The reason why CuDnn provider outputs NaNs is that it tries to access memory that is outside bounds of the input to the BN node. The MxNet exporter seems to have a bug wherein it *means* spatial == 1 but incorrectly stamps BatchNorm nodes with spatial == 0 and this leads to incorrect assumptions regarding input data shapes and illegal memory is accessed.

Please wait for the updated model and try again. (You can try updating the model yourself by changing the ModelProto so that the BN nodes are staped with spatial == 1).

Closing this issue as this is not an ORT issue.
